executor/read/mod.rs

```diff
@@ -62,35 +62,106 @@ pub(super) fn create_executors_for_pipeline(
     Ok(PatternExecutor::new(executors))
 }
 
+#[derive(Debug)]
 pub(crate) enum SuspendPoint {
     TabledCall(TabledCallSuspension),
     Nested(NestedSuspension),
 }
 
 impl SuspendPoint {
-    fn new_tabled_call(
-        executor_index: ExecutorIndex,
-        next_table_row: TableIndex,
-        input_row: MaybeOwnedRow<'static>,
-    ) -> Self {
-        Self::TabledCall(TabledCallSuspension { executor_index, next_table_row, input_row })
-    }
-
-    fn new_nested(executor_index: ExecutorIndex, branch_index: BranchIndex, input_row: MaybeOwnedRow<'static>) -> Self {
-        Self::Nested(NestedSuspension { executor_index, branch_index, input_row })
+    fn depth(&self) -> usize {
+        match self {
+            SuspendPoint::TabledCall(tabled_call) => tabled_call.depth,
+            SuspendPoint::Nested(nested) => nested.depth,
+        }
     }
 }
 
 #[derive(Debug)]
 pub(super) struct TabledCallSuspension {
     pub(crate) executor_index: ExecutorIndex,
+    pub(crate) depth: usize,
     pub(crate) input_row: MaybeOwnedRow<'static>,
     pub(crate) next_table_row: TableIndex,
 }
 
 #[derive(Debug)]
 pub(super) struct NestedSuspension {
     pub(crate) executor_index: ExecutorIndex,
+    pub(crate) depth: usize,
     pub(crate) branch_index: BranchIndex,
     pub(crate) input_row: MaybeOwnedRow<'static>,
 }
+
+#[derive(Debug)]
+pub(super) struct SuspendPointContext {
+    at_depth: usize,
+    suspended_points: Vec<SuspendPoint>,
+    restore_from: Vec<SuspendPoint>, //Peekable<std::vec::IntoIter<SuspendPoint>>,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+struct SuspendPointTrackerState(usize);
```

Comment:
So this is basically the number of suspend points? Can we just call it `SuspendPointCount`? Then when we `record_nested_pattern_entry` we get back a `SuspendPointCount` which I think makes sense?